### PR TITLE
[MKLDNN]Dynamic shape for ctc_greedy_decoder & ctc_greedy_decoder_seq…

### PR DESCRIPTION
**Pull request by zhangYiIntel**
### Details:
Support Dynamic Shape
 - *CtcGreedyDecoder*
 - *CtcGreedyDecoderSeqLen*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_node.cpp [hunk 1]
```diff
@@ -42,8 +38,8 @@ MKLDNNCTCGreedyDecoderNode::MKLDNNCTCGreedyDecoderNode(const std::shared_ptr<ngr
     if (getOriginalOutputsNumber() != 1)
         IE_THROW() << errorPrefix << "has invalid number of outputs edges: " << getOriginalOutputsNumber();
 
-    if (op->get_input_shape(DATA_INDEX)[0] != op->get_input_shape(SEQUENCE_LENGTH_INDEX)[0] &&
-        op->get_input_shape(DATA_INDEX)[1] != op->get_input_shape(SEQUENCE_LENGTH_INDEX)[1])
+    if (!op->get_input_partial_shape(DATA_INDEX)[0].compatible(op->get_input_partial_shape(SEQUENCE_LENGTH_INDEX)[0]) ||
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_node.cpp:41]: use `getInputShapeAtPort/getOutputShapeAtPort` instead

`const auto& dataDims = getInputShapeAtPort(DATA_INDEX).getDims();`
`const auto& seqDims = getInputShapeAtPort(SEQUENCE_LENGTH_INDEX).getDims();`
and check
`dimsEqualWeak(dataDims[0], seqDims[0])` etc

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_node.h [hunk 2]
```diff
@@ -15,12 +15,13 @@ class MKLDNNCTCGreedyDecoderNode : public MKLDNNNode {
 
     void getSupportedDescriptors() override {};
     void initSupportedPrimitiveDescriptors() override;
-    void createPrimitive() override {};
+    void createPrimitive() override;
     void execute(mkldnn::stream strm) override;
     bool created() const override;
+    void executeDynamicImpl(dnnl::stream strm) override;
+    void prepareParams() override;
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_node.h:22]: override `needPrepareParams` for the function to return false
instead `prepareParams`

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_seq_len_node.h [hunk 3]
```diff
@@ -15,9 +15,11 @@ class MKLDNNCTCGreedyDecoderSeqLenNode : public MKLDNNNode {
 
     void getSupportedDescriptors() override {};
     void initSupportedPrimitiveDescriptors() override;
-    void createPrimitive() override {};
+    void createPrimitive() override;
     void execute(mkldnn::stream strm) override;
     bool created() const override;
+    void executeDynamicImpl(dnnl::stream strm) override;
+    void prepareParams() override;
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_seq_len_node.h:22]: same as for CTCGreedyDecoder

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_seq_len_node.cpp [hunk 4]
```diff
@@ -42,7 +38,7 @@ MKLDNNCTCGreedyDecoderSeqLenNode::MKLDNNCTCGreedyDecoderSeqLenNode(const std::sh
     if (getOriginalOutputsNumber() != 2)
         IE_THROW() << errorPrefix << "has invalid number of outputs edges: " << getOriginalOutputsNumber();
 
-    if (op->get_input_shape(DATA_INDEX)[0] != op->get_input_shape(SEQUENCE_LENGTH_INDEX)[0])
+    if (!op->get_input_partial_shape(DATA_INDEX)[0].compatible(op->get_input_partial_shape(SEQUENCE_LENGTH_INDEX)[0]))
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_seq_len_node.cpp:41]: same as for CTCGreedyDecoder

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_seq_len_node.cpp [hunk 5]
```diff
@@ -170,4 +166,17 @@ bool MKLDNNCTCGreedyDecoderSeqLenNode::created() const {
     return getType() == CTCGreedyDecoderSeqLen;
 }
 
+void MKLDNNCTCGreedyDecoderSeqLenNode::createPrimitive() {
+    if (inputShapesDefined()) {
+        prepareParams();
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_seq_len_node.cpp:171]: in fact you no need to call `prepareParams()` here and in `CTCGreedyDecoder`

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            }
+
+            inputs.insert({funcInput.get_node_shared_ptr(), tensor});
+        }
+    }
+};
+
+TEST_P(CTCGreedyDecoderLayerCPUTest, CompareWithRefs) {
+    SKIP_IF_CURRENT_TEST_IS_DISABLED();
+    run();
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp:129]: also add CheckPluginRelatedResults(executableNetwork, "CTCGreedyDecoder");

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+using namespace CPUTestUtils;
+using namespace ov::test;
+
+namespace CPULayerTestsDefinitions {
+
+using TargetShapeParams = std::tuple<size_t,   // Sequence length T
+                                     size_t,   // Batch size N
+                                     size_t>;  // Number of classes
+
+using InputShapeParams = std::pair<std::vector<ov::PartialShape>,    // bounds for input dynamic shape
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp:27]: maybe better to pass PartialShape for `T` `N` `C` not for each inputs?

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    run();
+}
+
+namespace {
+// Common params
+// ElementType::f16 is not support by CPU Plugin yet
+const std::vector<ElementType> netPrecisions = {ElementType::f32};
+const std::vector<bool> mergeRepeated{true, false};
+const std::vector<InputShapeParams> inputShapesCTCDecoder = {
+    {{ov::PartialShape{-1, -1, -1}, ov::PartialShape{-1, -1}},
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp:138]: add tests cases with defined lower/upper bounds

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp [hunk 9]
```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2018-2021 Intel Corporation
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp:1]: the same comments as for `CtcGreedyDecoder`

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp (hunk 7 above)
Comment (zhangYiIntel) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp:27]: Hi yes, currently I just pass `T`, `N`, `C` as the `TargetShapeParams` for constructing `data` & `seq_len` shapes, I don't pass input shapes for `data` and `seq_len` separately. Maybe the naming `TargetShapeParams` is a bit misleading and I will change it to proper one. Thanks

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp (hunk 6 above)
Comment (zhangYiIntel) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp:129]: CheckPluginRelatedResults() method currently doesn't support `ov::runtime::ExecutableNetwork` API yet. will add a *TODO* here for reminder

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_node.cpp (hunk 1 above)
Comment (zhangYiIntel) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_ctc_greedy_decoder_node.cpp]: fix in 77986d46e02a2d3d8a43f7f35c8b76f777d0a6a4

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp (hunk 6 above)
Comment (mandrono) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/ctc_greedy_decoder.cpp:136]: after merge this PR it supported https://github.com/openvinotoolkit/openvino/pull/8339
uncomment please

**Review by mandrono — APPROVED** [review]
Please don't forget uncomment `CheckPluginRelatedResults`.
In other LGTM!

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**